File: src/py/netster.py
#!/usr/bin/env python3

import argparse
import logging as log

# Import the assignment modules.
# These imports can be specialized as necessary.
from a2 import *
from a3 import *
from a4 import *


DEFAULT_PORT=12345

# If we are a server, launch the appropriate methods to handle server
# functionality based on the input arguments.
# NOTE: The arguments should be extended with a custom dict or **kwargs
def run_server(host, port):
    log.info("Hello, I am a server...goodbye!")

# If we are a client, launch the appropriate methods to handle client
# functionality based on the input arguments
# NOTE: The arguments should be extended with a custom dict or **kwargs
def run_client(host, port):
    log.info("Hello, I am a client...goodbye!")

def main():
    parser = argparse.ArgumentParser(description="SICE Network netster")
    parser.add_argument('-p', '--port', type=str, default=DEFAULT_PORT,
                        help='listen on/connect to port <port> (default={}'
                        .format(DEFAULT_PORT))
    parser.add_argument('-i', '--iface', type=str, default='0.0.0.0',
                        help='listen on interface <dev>')
    parser.add_argument('-f', '--file', type=str,
                        help='file to read/write')
    parser.add_argument('-u', '--udp', action='store_true',
                        help='use UDP (default TCP)')
    parser.add_argument('-r', '--rudp', type=int, default=0,
                        help='use RUDP (1=stopwait, 2=gobackN)')
    parser.add_argument('-m', '--mcast', type=str, default='226.0.0.1',
                        help='use multicast with specified group address')
    parser.add_argument('-v', '--verbose', action='store_true',
                        help='Produce verbose output')
    parser.add_argument('host', metavar='host', type=str, nargs='?',
                        help='connect to server at <host>')

    args = parser.parse_args()

    # configure logging level based on verbose arg
    level = log.DEBUG if args.verbose else log.INFO

    f = None
    # open the file if specified
    if args.file:
        try:
            mode = "rb" if args.host else "wb"
            f = open(args.file, mode)
        except Exception as e:
            print("Could not open file: {}".format(e))
            exit(1)

    # Here we determine if we are a client or a server depending
    # on the presence of a "host" argument.
    if args.host:
        log.basicConfig(format='%(levelname)s:client: %(message)s',
                        level=level)
        run_client(args.host, args.port)
    else:
        log.basicConfig(format='%(levelname)s:server: %(message)s',
                        level=level)
        run_server(args.host, args.port)

    if args.file:
        f.close()
        
if __name__ == "__main__":
    main()

File: src/c/netster.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#include "a2.h"
#include "a3.h"
#include "a4.h"

#define DEFAULT_PORT 12345

typedef struct cfg {
  long  port;
  char *iface;
  char *server;
  char *mcast;
  FILE *fstream;
  int   use_udp;
  int   use_rudp;
} cfg_t;

/*
 * If we are a server, launch the appropriate methods to handle server
 * functionality based on the configuration arguments.
 */
void run_server(cfg_t *cfg) {
  fprintf(stderr, "Hello, I'm a server with no purpose...goodbye\n");  
  /* For example, if cfg->udp is true, then call the server UDP handler */
}

/*
 * If we are a client, launch the appropriate methods to handle client
 * functionality based on the configuration arguments.
 */
void run_client(cfg_t *cfg) {
  fprintf(stderr, "Hello, I'm a client with no purpose...goodbye\n");
  /* For example, if cfg->udp is true, then call the client UDP handler */
}

void usage(const char *argv0) {
  printf("Usage:\n");
  printf("  %s            start a server and wait for connection\n", argv0);
  printf("  %s <host>     connect to server at <host>\n", argv0);
  printf("\n");
  printf("Options:\n");
  printf("  -p, --port=<port>      listen on/connect to port <port>\n");
  printf("  -i, --iface=<dev>      listen on interface <dev>\n");
  printf("  -f, --file=<filename>  file to read/write\n");
  printf("  -u, --udp              use UDP (default TCP)\n");
  printf("  -r, --rudp             use RUDP (1=stopwait, 2=gobackN)\n");
  printf("  -m, --mcast=<addr>     multicast address\n");
}

int main(int argc, char **argv) {
  int c;
  char *fname = NULL;
  
  /* The configuration structure for netster */
  cfg_t cfg = {
    .port = DEFAULT_PORT,  /* which port to use */
    .iface = NULL,         /* which interface to use */
    .server = NULL,        /* the server to connect to */
    .mcast = NULL,         /* the multicast group to use */
    .fstream = NULL,       /* the open file stream if fname given */
    .use_udp = 0,          /* should we use UDP (default TCP) */
    .use_rudp = 0,         /* should we use RUDP (default TCP) */
  };
  
  static struct option long_options[] = {
    { .name = "port",           .has_arg = 1, .val = 'p' },
    { .name = "iface",          .has_arg = 1, .val = 'i' },
    { .name = "mcast",          .has_arg = 1, .val = 'm' },
    { .name = "file",           .has_arg = 1, .val = 'f' },
    { .name = "udp",            .has_arg = 0, .val = 'u' },
    { .name = "rudp",           .has_arg = 1, .val = 'r' },
    { 0 }
  };

  while (1) {
    c = getopt_long(argc, argv, "p:i:m:f:r:u", long_options, NULL);
    if (c == -1)
      break;
    
    switch (c) {
    case 'p':
      cfg.port = strtol(optarg, NULL, 0);
      if (cfg.port < 0 || cfg.port > 65535) {
	usage(argv[0]);
	return 1;
      }
      break;
      
    case 'i':
      cfg.iface = strdup(optarg);
      break;
      
    case 'u':
      cfg.use_udp = 1;
      break;
      
    case 'r':
      cfg.use_rudp = atoi(optarg);
      break;

    case 'm':
      cfg.mcast = strdup(optarg);
      break;

    case 'f':
      fname = strdup(optarg);
      break;
      
    default:
      usage(argv[0]);
      return 1;
      break;
    }
  }
  
  if (optind == argc - 1)
    cfg.server = strdup(argv[optind]);

  /* open the file if specified */
  if (fname) {
    const char *mode = (cfg.server) ? "r" : "w+";
    cfg.fstream = fopen(fname, mode);
    if (!cfg.fstream) {
      perror("fopen: ");
      exit(1);
    }
  }
  
  /* Here we decide if we started as a server or client! */
  if (cfg.server)
    run_client(&cfg);
  else
    run_server(&cfg);

  if (fname && cfg.fstream) {
    fclose(cfg.fstream);
  }
  
  return 0;
}
